File: src/services/commands/hubCommands.ts
import { invoke } from "@tauri-apps/api/core";

// ── Types ──

export interface HubAuthStatus {
  logged_in: boolean;
  user_id: string | null;
  email: string | null;
  display_name: string | null;
}

export interface HubUserInfo {
  id: string;
  email: string;
  display_name: string;
  peer_id: string | null;
}

export interface SharedAgent {
  id: string;
  user_id: string;
  display_name: string;
  name: string;
  description: string;
  original_agent_id: string | null;
  skills_count: number;
  notes_count: number;
  created_at: string;
  updated_at: string;
}

export interface SharedSkill {
  id: string;
  user_id: string;
  display_name: string;
  agent_id: string | null;
  agent_name: string | null;
  skill_name: string;
  description: string;
  body: string;
  created_at: string;
}

export interface SharedNote {
  id: string;
  user_id: string;
  display_name: string;
  agent_id: string | null;
  agent_name: string | null;
  title: string;
  note_type: string;
  tags: string[];
  body: string;
  created_at: string;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  limit: number;
  offset: number;
}

// ── Auth commands ──

export async function hubRegister(email: string, password: string, displayName?: string): Promise<HubAuthStatus> {
  return invoke("hub_register", { email, password, display_name: displayName });
}

export async function hubLogin(email: string, password: string): Promise<HubAuthStatus> {
  return invoke("hub_login", { email, password });
}

export async function hubLogout(): Promise<void> {
  return invoke("hub_logout");
}

export async function hubGetAuthStatus(): Promise<HubAuthStatus> {
  return invoke("hub_get_auth_status");
}

export async function hubGetMe(): Promise<HubUserInfo> {
  return invoke("hub_get_me");
}

export async function hubUpdateMe(displayName?: string, peerId?: string): Promise<HubUserInfo> {
  return invoke("hub_update_me", { display_name: displayName, peer_id: peerId });
}

// ── Share commands ──

export async function hubShareAgent(name: string, description: string, originalAgentId?: string): Promise<SharedAgent> {
  return invoke("hub_share_agent", { name, description, original_agent_id: originalAgentId });
}

export interface ShareSkillItem {
  name: string;
  description: string;
  body: string;
}

export async function hubShareSkills(agentId: string | null, skills: ShareSkillItem[]): Promise<SharedSkill[]> {
  return invoke("hub_share_skills", { agent_id: agentId, skills });
}

export interface ShareNoteItem {
  title: string;
  note_type: string;
  tags: string[];
  body: string;
}

export async function hubShareNotes(agentId: string | null, notes: ShareNoteItem[]): Promise<SharedNote[]> {
  return invoke("hub_share_notes", { agent_id: agentId, notes });
}

// ── List commands ──

export async function hubListAgents(q?: string, limit?: number, offset?: number, userId?: string): Promise<PaginatedResponse<SharedAgent>> {
  return invoke("hub_list_agents", { q, user_id: userId, limit, offset });
}

export async function hubListSkills(q?: string, agentId?: string, limit?: number, offset?: number, userId?: string): Promise<PaginatedResponse<SharedSkill>> {
  return invoke("hub_list_skills", { q, agent_id: agentId, user_id: userId, limit, offset });
}

export async function hubListNotes(q?: string, agentId?: string, limit?: number, offset?: number, userId?: string): Promise<PaginatedResponse<SharedNote>> {
  return invoke("hub_list_notes", { q, agent_id: agentId, user_id: userId, limit, offset });
}

// ── Delete commands ──

export async function hubDeleteAgent(id: string): Promise<void> {
  return invoke("hub_delete_agent", { id });
}

export async function hubDeleteSkill(id: string): Promise<void> {
  return invoke("hub_delete_skill", { id });
}

export async function hubDeleteNote(id: string): Promise<void> {
  return invoke("hub_delete_note", { id });
}

File: src-tauri/src/commands/hub_commands.rs
//! Tauri commands for Community Hub — register, login, share, browse, delete.

use crate::error::AppError;
use crate::services::hub_client::{self, HubClient};
use crate::settings::AppSettings;
use serde::Serialize;
use tauri::State;
use wa_shared::community::UserInfo;

// ── Return types ────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
pub struct HubAuthStatus {
    pub logged_in: bool,
    pub user_id: Option<String>,
    pub email: Option<String>,
    pub display_name: Option<String>,
}

// Re-export UserInfo as HubUserInfo for clarity at command boundary.
pub type HubUserInfo = UserInfo;

// ── Helpers ─────────────────────────────────────────────

fn make_hub_client(settings: &AppSettings) -> Result<HubClient, AppError> {
    let relay_url = settings.get().relay_url;
    Ok(HubClient::new(&relay_url, None))
}

fn make_authed_hub_client(
    app: &tauri::AppHandle,
    settings: &AppSettings,
) -> Result<HubClient, AppError> {
    let relay_url = settings.get().relay_url;
    let auth = hub_client::load_hub_auth(app)
        .ok_or_else(|| AppError::Api("Not logged in to Community Hub".into()))?;
    Ok(HubClient::new(&relay_url, Some(auth.token)))
}

// ── Auth commands ───────────────────────────────────────

#[tauri::command]
pub async fn hub_register(
    app: tauri::AppHandle,
    settings: State<'_, AppSettings>,
    email: String,
    password: String,
    display_name: Option<String>,
) -> Result<HubAuthStatus, AppError> {
    let client = make_hub_client(&settings)?;
    let resp = client
        .register(&email, &password, display_name.as_deref())
        .await?;

    hub_client::save_hub_auth(
        &app,
        &resp.token,
        &resp.user.id,
        &resp.user.email,
        &resp.user.display_name,
    )?;

    Ok(HubAuthStatus {
        logged_in: true,
        user_id: Some(resp.user.id),
        email: Some(resp.user.email),
        display_name: Some(resp.user.display_name),
    })
}

#[tauri::command]
pub async fn hub_login(
    app: tauri::AppHandle,
    settings: State<'_, AppSettings>,
    email: String,
    password: String,
) -> Result<HubAuthStatus, AppError> {
    let client = make_hub_client(&settings)?;
    let resp = client.login(&email, &password).await?;

    hub_client::save_hub_auth(
        &app,
        &resp.token,
        &resp.user.id,
        &resp.user.email,
        &resp.user.display_name,
    )?;

    Ok(HubAuthStatus {
        logged_in: true,
        user_id: Some(resp.user.id),
        email: Some(resp.user.email),
        display_name: Some(resp.user.display_name),
    })
}

#[tauri::command]
pub fn hub_logout(app: tauri::AppHandle) -> Result<(), AppError> {
    // Server uses stateless JWT — no server-side revocation needed.
    hub_client::clear_hub_auth(&app)
}

#[tauri::command]
pub fn hub_get_auth_status(app: tauri::AppHandle) -> Result<HubAuthStatus, AppError> {
    match hub_client::load_hub_auth(&app) {
        Some(auth) => Ok(HubAuthStatus {
            logged_in: true,
            user_id: Some(auth.user_id),
            email: Some(auth.email),
            display_name: Some(auth.display_name),
        }),
        None => Ok(HubAuthStatus {
            logged_in: false,
            user_id: None,
            email: None,
            display_name: None,
        }),
    }
}

// ── Profile commands ────────────────────────────────────

#[tauri::command]
pub async fn hub_get_me(
    app: tauri::AppHandle,
    settings: State<'_, AppSettings>,
) -> Result<HubUserInfo, AppError> {
    let client = make_authed_hub_client(&app, &settings)?;
    client.get_me().await
}

#[tauri::command]
pub async fn hub_update_me(
    app: tauri::AppHandle,
    settings: State<'_, AppSettings>,
    display_name: Option<String>,
    peer_id: Option<String>,
) -> Result<HubUserInfo, AppError> {
    let client = make_authed_hub_client(&app, &settings)?;
    let info = client
        .update_me(display_name.as_deref(), peer_id.as_deref())
        .await?;

    // Keep local auth store in sync with all profile fields
    if let Some(auth) = hub_client::load_hub_auth(&app) {
        hub_client::save_hub_auth(&app, &auth.token, &info.id, &info.email, &info.display_name)?;
    }

    Ok(info)
}

// ── Share commands ──────────────────────────────────────

#[tauri::command]
pub async fn hub_share_agent(
    app: tauri::AppHandle,
    settings: State<'_, AppSettings>,
    name: String,
    description: String,
    original_agent_id: Option<String>,
) -> Result<wa_shared::community::SharedAgent, AppError> {
    let client = make_authed_hub_client(&app, &settings)?;
    client
        .share_agent(&name, &description, original_agent_id.as_deref())
        .await
}

#[tauri::command]
pub async fn hub_share_skills(
    app: tauri::AppHandle,
    settings: State<'_, AppSettings>,
    agent_id: Option<String>,
    skills: Vec<wa_shared::community::ShareSkillItem>,
) -> Result<Vec<wa_shared::community::SharedSkill>, AppError> {
    let client = make_authed_hub_client(&app, &settings)?;
    client.share_skills(agent_id.as_deref(), skills).await
}

#[tauri::command]
pub async fn hub_share_notes(
    app: tauri::AppHandle,
    settings: State<'_, AppSettings>,
    agent_id: Option<String>,
    notes: Vec<wa_shared::community::ShareNoteItem>,
) -> Result<Vec<wa_shared::community::SharedNote>, AppError> {
    let client = make_authed_hub_client(&app, &settings)?;
    client.share_notes(agent_id.as_deref(), notes).await
}

// ── List commands ───────────────────────────────────────

#[tauri::command]
pub async fn hub_list_agents(
    _app: tauri::AppHandle,
    settings: State<'_, AppSettings>,
    q: Option<String>,
    user_id: Option<String>,
    limit: Option<u32>,
    offset: Option<u32>,
) -> Result<wa_shared::community::PaginatedResponse<wa_shared::community::SharedAgent>, AppError> {
    let client = make_hub_client(&settings)?;
    client.list_agents(q.as_deref(), user_id.as_deref(), limit, offset).await
}

#[tauri::command]
pub async fn hub_list_skills(
    _app: tauri::AppHandle,
    settings: State<'_, AppSettings>,
    q: Option<String>,
    agent_id: Option<String>,
    user_id: Option<String>,
    limit: Option<u32>,
    offset: Option<u32>,
) -> Result<wa_shared::community::PaginatedResponse<wa_shared::community::SharedSkill>, AppError> {
    let client = make_hub_client(&settings)?;
    client
        .list_skills(q.as_deref(), agent_id.as_deref(), user_id.as_deref(), limit, offset)
        .await
}

#[tauri::command]
pub async fn hub_list_notes(
    _app: tauri::AppHandle,
    settings: State<'_, AppSettings>,
    q: Option<String>,
    agent_id: Option<String>,
    user_id: Option<String>,
    limit: Option<u32>,
    offset: Option<u32>,
) -> Result<wa_shared::community::PaginatedResponse<wa_shared::community::SharedNote>, AppError> {
    let client = make_hub_client(&settings)?;
    client
        .list_notes(q.as_deref(), agent_id.as_deref(), user_id.as_deref(), limit, offset)
        .await
}

// ── Delete commands ─────────────────────────────────────

#[tauri::command]
pub async fn hub_delete_agent(
    app: tauri::AppHandle,
    settings: State<'_, AppSettings>,
    id: String,
) -> Result<(), AppError> {
    let client = make_authed_hub_client(&app, &settings)?;
    client.delete_agent(&id).await
}

#[tauri::command]
pub async fn hub_delete_skill(
    app: tauri::AppHandle,
    settings: State<'_, AppSettings>,
    id: String,
) -> Result<(), AppError> {
    let client = make_authed_hub_client(&app, &settings)?;
    client.delete_skill(&id).await
}

#[tauri::command]
pub async fn hub_delete_note(
    app: tauri::AppHandle,
    settings: State<'_, AppSettings>,
    id: String,
) -> Result<(), AppError> {
    let client = make_authed_hub_client(&app, &settings)?;
    client.delete_note(&id).await
}
